crates/federated-server/src/server/engine_reloader.rs

```diff
@@ -0,0 +1,143 @@
+use std::{path::PathBuf, sync::Arc};
+
+use engine::Engine;
+use futures_lite::{pin, StreamExt};
+use runtime_local::HooksWasi;
+use tokio::{
+    sync::{mpsc, watch},
+    task::JoinHandle,
+};
+use tokio_stream::{wrappers::WatchStream, Stream};
+
+use crate::server::gateway;
+
+use super::{
+    gateway::{EngineSender, EngineWatcher, GatewayRuntime, GraphDefinition},
+    graph_fetch_method::GraphStream,
+};
+
+/// Handles graph and config updates by constructing a new engine
+pub(super) struct EngineReloader {
+    engine_watcher: EngineWatcher,
+}
+
+pub(crate) type GraphSender = mpsc::Sender<GraphDefinition>;
+
+enum Update {
+    Graph(GraphDefinition),
+    Config(gateway_config::Config),
+}
+
+impl EngineReloader {
+    pub async fn spawn(
+        gateway_config: watch::Receiver<gateway_config::Config>,
+        mut graph_stream: GraphStream,
+        hot_reload_config_path: Option<PathBuf>,
```

Comment:
A bit confused why a `hot_reload_config_path` needs to be propagated down when `EngineReloader` seems like the struct that should handle all reloads to be honest. But seems like it was like this before so well. 🤷

Reply:
Yeah, weird isn't it.  Essentially there are two hot reloading mechanisms at the moment: 

1. The one used in the gateway to hot reload rate limiting stuff.  That's what this `hot_reload_config_path` is used for.
2. The one used by `grafbase dev` which is powered by the `watch::Receiver` here.

Seems like it should be cleaned up sometime.  But I've already spent a while on this PR and want to get onto other things, so a problem for the future.

Reply:
might be worth adding a quick comment here IMHO with exactly what you said.
thanks for the explanation!